docs/containerization-design.rst

```diff
@@ -0,0 +1,358 @@
+   as shown in the picture.
+
+#. Each Jobtype that is included in the flow will correspond to an init container that gets
+   initiated. This init conatainer will take the layer for the jobtype binary and add it to the volume
+   for the application container.
+
+Flow Container
+--------------
+
+1. A new class: "FlowContainer" will be created by refactoring code from the FlowRunnerManager. The purpose of this
+   class is to provide the anchor that initiates the flow orchestration as well as respond to control/health check
+   commands.
+
+#. The FlowContainer class is a simplified version of FlowRunnerManager with certain assumptions:
+
+   * This class will handle a single flow. Hence, the threading model can be simplified.
+   * There is no need to host polling logic as k8s based dispatch is done on the web server.
+   * There is no need to clean up execution directory or cache as the pod will be destroyed after the flow finishes.
+   * The above mentioned simplifications will have the effect of reducing the tech debt in flow orchestration.
+
+#. The web server needs to talk to the Kubernetes pods as the executor server hosts an AJAX API
+   endpoint for various control operations such as Cancel, Pause, Resume, FetchLogs etc. For the web server
+   to continue using this API endpoint, we need to enable communication between the Webserver (Which is outside
+   the k8s cluster) and the flow container pods. For this reason, we plan to use the
+   Ambassador Ingress Controller [2]_ between the Web Server and the Flow Container Pods.
+   More regarding the ingress controller `here <#ingress-controller>`_
+
+#. In the long-run, we do plan to bring in web server into Kubernetes as well, thereby eliminating the
+   Ingress Controller. For the short-term, we will continue to live with the added complexity.
+
+#. At Linkedin our internal analysis shows that APIs beyond Cancel, FetchLogs and Ping are rarely used. For
+   the sake of simplicity, for now FlowContainer endpoint will not support other AJAX APIs that are
+   supported on non-containerized version.
+
+#. During flow execution, flow and job life cycle events may need to be sent to Kafka through the
+   Event Reporter plugins as well as job/flow status updates may need to be made in Mysql db.
+   For sending events to Kafka, azkaban-exec-server’s cert issued by a valid certificate authority will be used
+   to authenticate flow containers. This and MySQL credentials will be pulled from Kubernetes secret.
+
+Ingress Controller
+------------------
+
+1. As mentioned in the [Flow Container Section](#flow-container), we will be utilizing the
+   Ambassador Ingress Controller [2]_ as a reverse proxy.
+
+#. The ingress controller will provide necessary routing between web server and the flow pods running on
+   kubernetes infrastructure. A key aspect of this architecture is that the routes between web server and flow pods
```

Comment:
Maybe add a sentence here: `Ambassador essentially provides a dynamically configurable reverse proxy that allows routing of request from the the web-server to the correct flow pod`.

Reply:
done